File: ndk2mec.pl
#junlysky.info@gmail.com
#!/usr/bin/env perl
use warnings;
use strict;

my $usage="perl $0 infile(.ndk) outfile ";
@ARGV > 0 or die "$usage \n";
my $infile=$ARGV[0];
my $outfile=$ARGV[1];
open(IN,"< $infile") or die "open $infile err :$!";
our @lines = <IN> ;
chomp(@lines);
close(IN);
open(OUT,"> $outfile") or die "err build $outfile:$!";
  print OUT "date time lon lat dep Mb Ms strike1 dip1 rake1 strike2 dip2 rake2  mrr mtt mpp mrt mrp mtp p_trend p_plu p_eigen b_trend b_plu b_eigen t_trend t_plu t_eigen scale exponent event_name \n";


our $i=0;
our $count=0;

our $date;
our $time;
our $lat; 
our $lon;
our $dep;
our $mag_mb;
our $mag_ms;
our $name;
our $time_tamp;
our $expond;
our $mrr;
our $mtt;
our $mpp;
our $mrt;
our $mrp;
our $mtp;
our $p_eigen;
our $p_plu;
our $p_trend;
our $b_eigen;
our $b_plu;
our $b_trend;
our $t_eigen;
our $t_plu;
our $t_trend;
our $scale;
our $str1;
our $dip1;
our $rake1;
our $str2;
our $dip2;
our $rake2;

our @line ;

while($i<@lines){
    @line = split /\s+/,$lines[$i];
    $count = $i%5 ;
#    printf "$count  @line \n";
    if($count == 0){
        $date = $line[1];
        $time = $line[2];
        $lat = $line[3];
        $lon = $line[4];
        $dep = $line[5];
        $mag_mb = $line[6];
        $mag_ms = $line[7];
    }
    elsif( $count ==1 ){
        $name = $line[0];
    }
    elsif( $count == 2 ){
         $time_tamp = $line[10];
 #       $time_tamp = substr($time_tamp,0,1);
    }
    elsif( $count == 3 ){
        $expond = $line[0];
        $mrr = $line[1];
        $mtt = $line[3];
        $mpp = $line[5];
        $mrt = $line[7];
        $mrp = $line[9];
        $mtp = $line[11];
 # r is up ,t is south, p is east; 
    }
    elsif( $count == 4){
        $p_eigen=$line[1];
        $p_plu=$line[2];
        $p_trend=$line[3];
        $b_eigen=$line[4];
        $b_plu=$line[5];
        $b_trend=$line[6];
        $t_eigen=$line[7];
        $t_plu=$line[8];
        $t_trend=$line[9];
        $scale=$line[10];
        $str1=$line[11];
        $dip1=$line[12];
        $rake1=$line[13];
        $str2=$line[14];
        $dip2=$line[15];
        $rake2=$line[16];

        print  OUT "$date $time $lon $lat $dep $mag_mb $mag_ms $str1 $dip1 $rake1 $str2 $dip2 $rake2 $ mrr $mtt $mpp $mrt $mrp $mtp $p_trend $p_plu $p_eigen $b_trend $b_plu $b_eigen $t_trend $t_plu $t_eigen $scale $expond $name \n";
    }
#    else die "something is err:$!";
    $i++;
}
close(OUT);

File: ndk2mec.c
//  junlysky.info@gmail.com
# include <stdio.h>
# include <stdlib.h>
# include <string.h>
# include <math.h>

enum hang { one, two, three, four, five };

int main(int argc,char **argv){
    FILE *fpin;
    FILE *fpout;
    
    if(argc<=2){
        printf("usage:ndk2mec infile outfile \n");
        return 0;
    }
    
    --argc;
    ++argv;
    fpin=fopen(*argv,"r");
    if(fpin == NULL){
        printf("error to open %s.\n",*argv);
        return 0;
    }
    
    ++argv;
    fpout=fopen(*argv,"w");
    if(fpout==NULL){
        printf("error to build %s.\n",*argv);
        return 0;
    }

    char lines[512];
 
    char catlog[4],date[10],time[10],gloca[24];
    float lat,lon,dep,mag_mb,mag_ms;
    char name[16],cmt_para[64];
    char lin3[80];
    int exponent;
    float mrr,mtt,mpp,mrt,mrp,mtp; //r is up,t is south,p is east
    float err1,err2,err3,err4,err5,err6;
    char ver[3];
    float p_gen,p_plu,p_azi,b_gen,b_plu,b_azi,t_gen,t_plu,t_azi,scalar;
    float str1,dip1,rak1,str2,dip2,rak2;

    fprintf (fpout, "date time lon lat depth Mb Ms strike1 dip1  rake1  strike2  dip2  rake2   mrr  mtt  mpp  mrt  mrp  mtp  p_azi  p_plu  p_eigen  b_azi  b_plu  b_eigen  t_azi  t_plu  t_eigen  scale  exponent  name \n");

    int i=0;
    enum hang count; 
    while(fgets(lines,sizeof(lines),fpin)){
        count=i%5;
      
        if(count == 0){
            sscanf(lines,"%s %s %s %f %f %f %f %f %s ",catlog,date,time,&lat,&lon,&dep,&mag_mb,&mag_ms,gloca);
        }
        else if(count ==1){
            sscanf(lines,"%s %s",name,cmt_para);
        }
        else if(count ==2 ){
            sscanf(lines,"%s",lin3);
        }
        else if(count == 3){
            sscanf(lines,"%d %f %f %f %f %f %f %f %f %f %f %f %f",&exponent,&mrr,&err1,&mtt,&err2,&mpp,&err3,&mrt,&err4,&mrp,&err5,&mtp,&err6);
        }
        else if (count ==4){
            sscanf(lines,"%s %f %f %f %f %f %f %f %f %f %f %f %f %f %f %f %f",ver,&p_gen,&p_plu,&p_azi,&b_gen,&b_plu,&b_azi,&t_gen,&t_plu,&t_azi,&scalar,&str1,&dip1,&rak1,&str2,&dip2,&rak2);
            fprintf(fpout,"%s %s %.2f %.2f %.1f %.1f %.1f %.0f %.0f %.0f %.0f %.0f %.0f %.3f %.3f %.3f %.3f %.3f %.3f %.1f %.1f %.3f %.1f %.1f %.3f %.1f %.1f %.3f %.3f %2d %s \n",date,time,lon,lat,dep,mag_mb,mag_ms,str1,dip1,rak1,str2,dip2,rak2,mrr,mtt,mpp,mrt,mrp,mtp,p_azi,p_plu,p_gen,b_azi,b_plu,b_gen,t_azi,t_plu,t_gen,scalar,exponent,name);
        }
        else {
             printf ("something is wrong! \n");
            return 0;
        }
        i++;
    }

    fclose(fpin);
    fclose(fpout);
    return 0;
}